### ISSUE TITLE
dd-trace integration: missing resource name

### ISSUE THREAD

**Issue by mwinstanley** [express-promise-router/express-promise-router#146]
We're using `express` and `express-promise-router` along with `dd-trace` to trace our Express requests with Datadog.

`express-promise-router` is creating [a wrapper function](https://github.com/express-promise-router/express-promise-router/blob/trunk/lib/express-promise-router.js#L51-L60) around the handler that is anonymous, meaning that it doesn't have a function name and loses the function name data from the original handler.

This results in a resource name of `<anonymous>` in `dd-trace` because there isn't a function name ([how dd-trace uses the function name](https://github.com/DataDog/dd-trace-js/blob/8eaaebc70e619fae91c97c23d925d019896508c7/packages/dd-trace/src/plugins/util/web.js#L106)), making it harder to debug problems.

Is it reasonable to modify `express-promise-router` to ensure that the wrapper functions have the same function name as the original handler function? Something like:
```
  var wrapperObj = {};
  if (handler.length === 4) {
    wrapperObj[handler.name] = function (err, req, res, next) {
      handleReturn([err, req, res, next]);
    };
  } else {
    wrapperObj[handler.name] = function (req, res, next) {
      handleReturn([req, res, next]);
    };
  }
  return wrapperObj[handler.name];
```

**Comment by mormahr** [comment]
I don't understand your proposed solution. How does this modify the function name? Wouldn't a named function have to be named by direct syntax (e.g. `function theName() {}` or `const theName = function () {}`)?

I just quickly tried the object assignment method and it assigned an empty string as the name.

```js
const obj = {};
const name = "testName";

obj[name] = function () { return true };

console.log(obj[name].name) // => ""
```

**Comment by mwinstanley** [comment]
Oops, in trying to clean up my solution, I made it not work. In order to assign a name like this, it has to be done at object creation time:
```
const name = "testName";

const obj = {
  [name]: function () { return true },
};

console.log(obj[name].name) // => "testName"
```

Or in the actual library:
```
  if (handler.length === 4) {
    var wrapperObj = {
      [handler.name]: function (err, req, res, next) {
        handleReturn([err, req, res, next]);
      },
    };
    return wrapperObj[handler.name];
  }
  var wrapperObj = {
    [handler.name]: function (req, res, next) {
      handleReturn([req, res, next]);
    },
  };
  return wrapperObj[handler.name];
```
See [Symbols as function names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name)

**Comment by mormahr** [comment]
Thanks for the clarification. I submitted a PR (#147) with your changes and some quick tests.

**Comment by synchrone** [comment]
the same issue affects newrelic's instrumentation that reports middlewares by their names to APM

**Comment by mormahr** [comment]
Released as v4.1.0
